luigi/mypy.py

```diff
@@ -0,0 +1,457 @@
+        """Apply all the necessary transformations to the underlying gokart.Task"""
+        info = self._cls.info
+        attributes = self.collect_attributes()
+
+        if attributes is None:
+            # Some definitions are not ready. We need another pass.
+            return False
+        for attr in attributes:
+            if attr.type is None:
+                return False
+        # If there are no attributes, it may be that the semantic analyzer has not
+        # processed them yet. In order to work around this, we can simply skip generating
+        # __init__ if there are no attributes, because if the user truly did not define any,
+        # then the object default __init__ with an empty signature will be present anyway.
+        if (
+            "__init__" not in info.names or info.names["__init__"].plugin_generated
+        ) and attributes:
+            args = [attr.to_argument(info, of="__init__") for attr in attributes]
+            add_method_to_class(
+                self._api, self._cls, "__init__", args=args, return_type=NoneType()
+            )
+        info.metadata[METADATA_TAG] = {
+            "attributes": [attr.serialize() for attr in attributes],
+        }
+
+        return True
+
+    def _get_assignment_statements_from_if_statement(
+        self, stmt: IfStmt
+    ) -> Iterator[AssignmentStmt]:
+        for body in stmt.body:
+            if not body.is_unreachable:
+                yield from self._get_assignment_statements_from_block(body)
+        if stmt.else_body is not None and not stmt.else_body.is_unreachable:
+            yield from self._get_assignment_statements_from_block(stmt.else_body)
+
+    def _get_assignment_statements_from_block(
+        self, block: Block
+    ) -> Iterator[AssignmentStmt]:
+        for stmt in block.body:
+            if isinstance(stmt, AssignmentStmt):
+                yield stmt
+            elif isinstance(stmt, IfStmt):
+                yield from self._get_assignment_statements_from_if_statement(stmt)
+
+    def collect_attributes(self) -> Optional[List[TaskAttribute]]:
+        """Collect all attributes declared in the task and its parents.
+
+        All assignments of the form
+
+          a: SomeType
+          b: SomeOtherType = ...
+
+        are collected.
+
+        Return None if some base class hasn't been processed
+        yet and thus we'll need to ask for another pass.
+        """
+        cls = self._cls
+
+        # First, collect attributes belonging to any class in the MRO, ignoring duplicates.
+        #
+        # We iterate through the MRO in reverse because attrs defined in the parent must appear
+        # earlier in the attributes list than attrs defined in the child.
+        #
+        # However, we also want attributes defined in the subtype to override ones defined
+        # in the parent. We can implement this via a dict without disrupting the attr order
+        # because dicts preserve insertion order in Python 3.7+.
```

Comment:
Luigi still advertises to be 3.6 compatible. Though, support for 3.6 should be removed (readme reference, tox reference, github actions test execution, etc.).

Reply:
I saw elsewhere in this PR that the mypy stuff requires 3.8 or later. What is the behavior when someone tries to use it with 3.6 or 3.7?

Reply:
> I saw elsewhere in this PR that the mypy stuff requires 3.8 or later
Sorry about your confusing.
I made a mistake to set a type annotations like `list[Any]` which supported after 3.8
I fixed this issue, the tests worked on 3.6 and 3.7.

I'll check how it works on Python 3.6 via tests.
If the test become flaky, one of workaround is ommit supporting mypy plugion on 3.6

Reply:
After inspecting issues about python 3.6 and 3.7, some built-in features are lacking (e.g. some type annotation features like `Final` and `Literal`).
 
Can I want to omit these versions for this plugin because this feature is not core required to use luigi?

Reply:
[4744560](https://github.com/spotify/luigi/pull/3315/commits/474456006375343ea3de5a45a2873789ecd0f646)

Reply:
Yeah, i'm fine with the included solution (requiring a minimum python version to import and use the mypy plugin).